woocommerce/woocommerce-ios
# [Mobile Payments] Charge actual order total

Closes #3937 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Please note there are several parts of the implementation that are not production ready.⚠️

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by charging the actual total amount of an order.

So far we have been capturing a payment with a hardcoded amount. This PR adds logic to charge the actual total order.

## Changes
* Update PaymentIntentParams to take an NSDecimal, instead of a UInt to model the charge amount. The reason for that is that a UInt is specific to Stripe. We already have mechanisms in place to format the order total amount as an NSDecimal, so it makes sense to do it here as well.
* Transform the amount to the currency’s smallest unit, [as required by Stripe](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPPaymentIntentParameters.html#/c:objc(cs)SCPPaymentIntentParameters(py)amount), in the internal method (internal to Hardware) that transform the intent parameters into what Stripe requires.

## Testing
Testing is a bit complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

Once the store is setup and there is a hardware reader available:
1. Checkout the branch. Run `bundle exec pod install`. 
2. Add a breakpoint in line 109 of CardPresentPaymentStore
<img width="350" alt="Screen Shot 2021-04-09 at 06 20 31" src="https://user-images.githubusercontent.com/2722505/114166424-b7098600-98fb-11eb-9908-1845459c2b59.png">

3. Build and run on a device. 
4. Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
5. Log in to an account that matches a store that has been configured for testing.
6. Navigate to Settings in the app. (Tap the gear button in the top right)
7. In the "Store Settings" section, tap "Manage Card Reader".
8. Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
9. Tap "Connect card reader"
10. Wait until the app suggests one reader to connect to.
11. Tap "Connect to reader"
12. Make sure the reader light turns solid blue.
13. Navigate to an order (any order with payment method Cash On Delivery).
14. Scroll down to the "Collect Payment" button
15. Tap the button. Follow the instructions on screen

When the breakpoint is triggered, `po intent` and check that the amount matches the expectation.
<img width="350" alt="Screen Shot 2021-04-09 at 06 23 07" src="https://user-images.githubusercontent.com/2722505/114166722-1a93b380-98fc-11eb-839d-741edd4db22e.png">

For example, for an order with a total of 210.00, the amount in the intent should read 21000



Update release notes:
